internal/models/factor.go

```diff
@@ -205,23 +206,36 @@ func DeleteUnverifiedFactors(tx *storage.Connection, user *User, factorType stri
 	return nil
 }
 
-func (f *Factor) CreateChallenge(ipAddress string) *Challenge {
+func (f *Factor) CreateChallenge(tx *storage.Connection, ipAddress string) *Challenge {
 	id := uuid.Must(uuid.NewV4())
 	challenge := &Challenge{
 		ID:        id,
 		FactorID:  f.ID,
 		IPAddress: ipAddress,
 	}
+
 	return challenge
 }
+func (f *Factor) WriteChallengeToDatabase(tx *storage.Connection, challenge *Challenge) error {
+	if challenge.FactorID != f.ID {
```

Comment:
This shouldn't happen since we check that factors and challenges are linked. It's there just to provide additional certainty